src/phoenix/utilities/ratelimits.py

```diff
@@ -0,0 +1,210 @@
+                self.spend_tokens_if_available(token_cost)
+                break
+            except UnavailableTokensError:
+                await asyncio.sleep(1 / self.rate)
+                continue
+
+    def effective_rate(self) -> Numeric:
+        return self.total_tokens / (time.time() - self.created)
+
+
+class LimitStore:
+    """
+    A singleton store for collections of rate limits grouped by service key.
+
+    LimitStore is a singleton because we want to share rate limits across all calls controlled by
+    phoenix's rate limiting mechanism.
+    """
+
+    _singleton = None
+
+    def __new__(cls) -> "LimitStore":
+        if not cls._singleton:
+            cls._singleton = super().__new__(cls)
+            cls._singleton._rate_limits = defaultdict(dict)
+        return cls._singleton
+
+    _rate_limits: Dict[str, Dict[str, TokenLimiter]]
+
+    def set_rate_limit(
+        self,
+        key: str,
+        limit_type: str,
+        per_minute_rate_limit: Numeric,
+        enforcement_window_minutes: Optional[int] = None,
+    ) -> None:
+        # default to 1 minute enforcement window
+        enforcement_window_minutes = (
+            enforcement_window_minutes if enforcement_window_minutes is not None else 1
+        )
+        max_tokens = per_minute_rate_limit * enforcement_window_minutes
+        if limits := self._rate_limits[key]:
+            if limit := limits.get(limit_type):
+                limit.refresh_limit(per_minute_rate_limit, max_tokens)
+                return
+        limits[limit_type] = TokenLimiter(
+            per_minute_rate_limit,
+            0,
+            max_tokens,
+        )
+
+    def get_rate_limits(self, key: str) -> Dict[str, TokenLimiter]:
+        return self._rate_limits[key]
+
+    def wait_for_rate_limits(self, key: str, rate_limit_costs: Dict[str, Numeric]) -> None:
+        rate_limits = self._rate_limits[key]
+        for limit_type, cost in rate_limit_costs.items():
+            if limit := rate_limits.get(limit_type):
+                limit.wait_for_then_spend_available_tokens(cost)
+
+    async def async_wait_for_rate_limits(
+        self, key: str, rate_limit_costs: Dict[str, Numeric]
+    ) -> None:
+        rate_limits = self._rate_limits[key]
+        for limit_type, cost in rate_limit_costs.items():
+            if limit := rate_limits.get(limit_type):
+                await limit.async_wait_for_then_spend_available_tokens(cost)
+
+    def spend_rate_limits(self, key: str, rate_limit_costs: Dict[str, Numeric]) -> None:
+        rate_limits = self._rate_limits[key]
+        for limit_type, cost in rate_limit_costs.items():
+            if limit := rate_limits.get(limit_type):
+                limit.spend_tokens(cost)
+
+
+class OpenAIRateLimiter:
+    def __init__(self, api_key: str) -> None:
+        self._store = LimitStore()
+        self._api_key = api_key
+
+    def key(self, model_name: str) -> str:
+        return f"openai:{self._api_key}:{model_name}"
```

Comment:
I would avoid keys that use the openai key - it's a security risk. We shouldn't even touch the key. The model name is already effectively unique per runtime.